File: src/main/java/io/github/jsnimda/common/gui/ConfigOptionHotkeyWidget.java
package io.github.jsnimda.common.gui;

import java.util.Arrays;
import java.util.List;

import com.mojang.blaze3d.platform.GlStateManager;

import io.github.jsnimda.common.input.IKeybind;
import org.lwjgl.glfw.GLFW;

import io.github.jsnimda.common.config.options.ConfigHotkey;
import io.github.jsnimda.common.input.GlobalInputHandler;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.Element;
import net.minecraft.client.gui.widget.ButtonWidget;
import net.minecraft.client.resource.language.I18n;
import net.minecraft.util.Identifier;

public class ConfigOptionHotkeyWidget extends ConfigOptionWidgetBase<ConfigHotkey> {

  private ButtonWidget setKeyButton = new ButtonWidget(10, 10, 200, 20, "", x -> {
    GlobalInputHandler.INSTANCE.setCurrentAssigningKeybind(configOption.getMainKeybind());
  });

  private static Identifier WIDGETS_TEXTURE = new Identifier("inventoryprofiles", "textures/gui/widgets.png");

  public boolean  isInConfigHotkeyOverlay = false;
  public IKeybind targetKeybind;

  protected ConfigOptionHotkeyWidget(ConfigHotkey configOption) {
    super(configOption);
    targetKeybind = configOption.getMainKeybind();
  }

  @Override
  public void render(int mouseX, int mouseY, float partialTicks) {
    super.render(mouseX, mouseY, partialTicks);

    int textureX = 20 + ((targetKeybind.isSettingsModified() || !configOption.getAlternativeKeybinds().isEmpty()) ? 20 : 0);
    int textureY = 160 + targetKeybind.getSettings().activateOn.ordinal() * 20;
    if (targetKeybind.getKeyCodes().isEmpty()) {
      textureY = 140;
    }

    MinecraftClient.getInstance().getTextureManager().bindTexture(WIDGETS_TEXTURE);
    GlStateManager.disableDepthTest();
    blit(this.x, this.y, textureX, textureY, 20, 20, 256, 256);
    GlStateManager.enableDepthTest();

    if (VHLine.contains(this.x, this.y, this.x + 20, this.y + 20, mouseX, mouseY)) {
      // show Advanced Keybind Settings
      Tooltips.getInstance().addTooltip(getKeybindSettingsTooltip(), mouseX, mouseY);
    }

    setKeyButton.x = x + 20 + 2;
    setKeyButton.y = y;
    setKeyButton.setWidth(availableWidth - 20 - 2);
    String displayText = targetKeybind.getDisplayText();
    setKeyButton.setMessage(GlobalInputHandler.INSTANCE.getCurrentAssigningKeybind() == targetKeybind
      ? ("> §e" + displayText + "§r <") : displayText);
    setKeyButton.render(mouseX, mouseY, partialTicks);

  }

  private static String textPrefix = "inventoryprofiles.common.gui.config.";
  private static String translate(String suffix) {
    return I18n.translate(textPrefix + suffix);
  }
  private String getKeybindSettingsTooltip() {
    String yes = translate("yes");
    String no = translate("no");
    String s = "§n" + translate("advanced_keybind_settings");
    s += String.format("\n%s: %s", translate("activate_on"),      "§9" + targetKeybind.getSettings().activateOn.toString());
    s += String.format("\n%s: %s", translate("context"),          "§9" + targetKeybind.getSettings().context.toString());
    s += String.format("\n%s: %s", translate("allow_extra_keys"), "§6" + (targetKeybind.getSettings().allowExtraKeys ? yes : no));
    s += String.format("\n%s: %s", translate("order_sensitive"),  "§6" + (targetKeybind.getSettings().orderSensitive ? yes : no));
    s += "\n\n" + translate("keybind_settings_tips");
    return s;
  }

  protected void onClickKeybindSettingsIcon() {
    if (!isInConfigHotkeyOverlay) {
      MinecraftClient.getInstance().openScreen(new ConfigOptionHotkeyOverlayScreen(configOption));
    }
  }

  @Override
  protected void reset() {
    targetKeybind.resetKeyCodesToDefault();
  }

  @Override
  protected boolean resetButtonActive() {
    return targetKeybind.isKeyCodesModified();
  }

  @Override
  public boolean mouseClicked(double d, double e, int i) {
    if (super.mouseClicked(d, e, i)) {
      return true;
    }
    if (VHLine.contains(this.x, this.y, this.x + 20, this.y + 20, (int)d, (int)e)) {
      if (i == GLFW.GLFW_MOUSE_BUTTON_RIGHT) {
        targetKeybind.resetSettingsToDefault();
      } else if (i == GLFW.GLFW_MOUSE_BUTTON_LEFT) {
        onClickKeybindSettingsIcon();
      }
      return true;
    }
    return false;
  }

  @Override
  public List<? extends Element> children() {
    return Arrays.asList(resetButton, setKeyButton);
  }

  
}

File: src/main/java/io/github/jsnimda/common/gui/widgets/ConfigOptionHotkeyWidget.kt
package io.github.jsnimda.common.gui.widgets

import io.github.jsnimda.common.config.options.ConfigHotkey
import io.github.jsnimda.common.gui.Tooltips
import io.github.jsnimda.common.gui.screen.ConfigOptionHotkeyDialog
import io.github.jsnimda.common.input.GlobalInputHandler
import io.github.jsnimda.common.input.IKeybind
import io.github.jsnimda.common.vanilla.VanillaUtils
import io.github.jsnimda.common.vanilla.alias.I18n
import io.github.jsnimda.common.vanilla.alias.Identifier
import io.github.jsnimda.common.vanilla.render.bindTexture
import io.github.jsnimda.common.vanilla.render.blit
import org.lwjgl.glfw.GLFW.GLFW_MOUSE_BUTTON_LEFT
import org.lwjgl.glfw.GLFW.GLFW_MOUSE_BUTTON_RIGHT

private val WIDGETS_TEXTURE =
  Identifier("inventoryprofiles", "textures/gui/widgets.png")
private const val textPrefix = "inventoryprofiles.common.gui.config."
private fun translate(suffix: String): String {
  return I18n.translate(textPrefix + suffix)
}

class ConfigOptionHotkeyWidget(configOption: ConfigHotkey) : ConfigOptionBaseWidget<ConfigHotkey>(configOption) {
  private val setKeyButton = ButtonWidget { -> GlobalInputHandler.currentAssigningKeybind = configOption.mainKeybind }
  private val iconButton = object : ButtonWidget({ button ->
    if (button == GLFW_MOUSE_BUTTON_RIGHT) {
      targetKeybind.resetSettingsToDefault()
    } else if (button == GLFW_MOUSE_BUTTON_LEFT) {
      onClickKeybindSettingsIcon()
    }
  }) {
    override fun renderButton(hovered: Boolean) {
      bindTexture(WIDGETS_TEXTURE)
//      disableDepthTest()
      val textureX = 20 + if (targetKeybind.isSettingsModified || !configOption.alternativeKeybinds.isEmpty()) 20 else 0
      val textureY = 160 + targetKeybind.settings.activateOn.ordinal * 20
      blit(screenX, screenY, textureX, textureY, 20, 20)
//      enableDepthTest()
    }
  }

  var targetKeybind: IKeybind = configOption.mainKeybind
  val keybindDisplayText
    get() = targetKeybind.displayText.let {
      if (GlobalInputHandler.currentAssigningKeybind === targetKeybind) "> §e$it§r <" else it
    }

  override fun render(mouseX: Int, mouseY: Int, partialTicks: Float) {
    setKeyButton.text = keybindDisplayText
    super.render(mouseX, mouseY, partialTicks)
    if (iconButton.contains(mouseX, mouseY)) { // show Advanced Keybind Settings
      Tooltips.addTooltip(keybindSettingsTooltipText, mouseX, mouseY)
    }
  }

  private val keybindSettingsTooltipText: String
    get() {
      val yes = translate("yes")
      val no = translate("no")
      return with(targetKeybind.settings) {
        """§n${translate("advanced_keybind_settings")}
          |${translate("activate_on")}: §9$activateOn
          |${translate("context")}: §9$context
          |${translate("allow_extra_keys")}: §6${if (allowExtraKeys) yes else no}
          |${translate("order_sensitive")}: §6${if (orderSensitive) yes else no}
          |
          |${translate("keybind_settings_tips")}""".trimMargin()
      }
    }

  protected fun onClickKeybindSettingsIcon() {
    VanillaUtils.openScreen(ConfigOptionHotkeyDialog(configOption))
  }

  override fun reset() {
    targetKeybind.resetKeyCodesToDefault()
  }

  override fun resetButtonActive(): Boolean {
    return targetKeybind.isKeyCodesModified
  }

  init {
    flow.least.add(iconButton, 20, false, 20)
    flow.least.addSpace(2)
    flow.addAndFit(setKeyButton)
  }
}